bitwarden_access_manager.py

```diff
@@ -0,0 +1,672 @@
+            email = data["email"],
+            type = data["type"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[member]]",
+            f"member_id = {self.id}",
+            f"member_name = {self.name}",
+            f"email = {self.email}",
+            f"type = {str(self.type)}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+class GroupMember(NamedTuple):
+    member_id: int
+    member_name: str
+    group_name: str
+
+    def get_id(self) -> str:
+        # Our generic differ has the ability to turn add/removes into changes
+        # for pairs with the same id, but this does not apply to memberships,
+        # which do not themselves have an id, so the identity to group on is
+        # the value itself.
+        return f"{self.member_id}@{self.group_name}"
+
+    def format_toml(self) -> str:
+        # Needed to satisfy Diffable, but not used in this case.
+        raise Exception(
+            "Group memberships are not expressed in toml, "
+            "please print the diffs in some other way."
+        )
+
+class Group(NamedTuple):
+    id: str
+    name: str
+    access_all: Optional[bool]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Group:
+        access_all: Optional[bool] = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+
+        return Group(
+            id=data["group_id"],
+            name=data["group_name"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[group]]",
+            f"group_id = {self.id}",
+            f"group_name = {self.name}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+
+class MemberCollectionAccess(NamedTuple):
+    name: str
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> MemberCollectionAccess:
+        return MemberCollectionAccess(
+            name=data["member_name"],
+        )
+
+    def format_toml(self) -> str:
+        return (
+            "{ member_name = "
+            + self.name
+            + '"}'
+        )
+
+class GroupCollectionAccess(NamedTuple):
+    name: str
+    access: str
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> GroupCollectionAccess:
+        return GroupCollectionAccess(
+            name=data["group_name"],
+            access=data["access"],
+        )
+
+    def format_toml(self) -> str:
+        return (
+            "{ group_name = "
+            + self.name
+            + '", access = "'
+            + self.access
+            + '" }'
+        )
+
+class Collection(NamedTuple):
+    id: str
+    external_id: str
```

Comment:
At the moment we are not using these external ids, I think we could omit them entirely.

Reply:
@ruuda ExternalID allow you to retrieve a friendly name from the collection.

When you query a collection using is ID the api don't return the name but the externalId. We should use it in order to store in the config a name that can be understand by humain and not just an id.

```
{
  "externalId": "external_id_123456",
  "object": "collection",
  "id": "539a36c5-e0d2-4cf9-979e-51ecf5cf6593",
  "groups": [
    {
      "id": "bfbc8338-e329-4dc0-b0c9-317c2ebf1a09",
      "readOnly": true
    }
  ]
}
```

Reply:
Hmm, does the API not return the name at all? And does every collection have a unique external id, even when we do not fill it in? In the UI there is this:

![2023-03-07-bitwarden-external-id](https://user-images.githubusercontent.com/506953/223485454-98e6fa4e-ab1b-4257-8f34-d451e9015388.png)

I thought this “external id” was for when you want to connect with e.g. Active Directory and map objects in Bitwarden to objects in your directory server, but we are not using that.

Reply:
Yes you're right the `external id` is used when using `Directoy Connector` to connect to a `directory service` like `AD`.

The API only return the collection ID, I was thinking to use the external id to retrieve their names. I don't see a better approach to this even if it doesn't really solve the problem.